cron/jobstats.go

```diff
@@ -0,0 +1,117 @@
+package cron
+
+import (
+	"database/sql"
+	"fmt"
+)
+
+type Job struct {
+	Category	string
+	Status		string
+	Count		int
+}
+
+//jobs/submitted
+func GetSubmittedJobCounts(db *sql.DB, startDay string, endDay string)([]Job, error) {
+	var jobType *string
+	var count int
+	//skip curl wrapper jobs
+	query := `SELECT b.job_type, count(b.*) AS count
+            FROM (
+              SELECT a.id,
+                CASE WHEN array_length(a.types, 1) = 1 THEN a.types[1]
+              ELSE 'Mixed'
+           END AS job_type
+           FROM (
+               SELECT j.id, array_agg(DISTINCT t.name) AS types FROM jobs j
+               JOIN job_steps s ON j.id = s.job_id
+               JOIN job_types t ON s.job_type_id = t.id
+               WHERE j.start_date >= ($1 :: DATE) AND j.start_date <= ($2 :: DATE)
+               AND j.app_id != '1e8f719b-0452-4d39-a2f3-8714793ee3e6'
+               GROUP BY j.id
+           ) AS a
+          ) AS b
+         GROUP BY b.job_type
+         ORDER BY b.job_type;`
+
+	rows, err := db.Query(query, startDay, endDay)
+	if err != nil {
+		return nil, err
+	}
+
+	defer rows.Close()
+
+	var jobs []Job
+
+	for i := 0; rows.Next(); i++ {
+		err := rows.Scan(&jobType, &count)
+		if err != nil {
+			return nil, err
+		}
+		jobs = append(jobs, Job{getStringValue(jobType), "Submitted", count})
+		output := fmt.Sprintf("Total no.of %[1]v jobs Submitted in last 24 hours: %[2]v", getStringValue(jobType), count)
+		fmt.Println(output)
+
+	}
+
+	err = rows.Err()
+	if err != nil {
+		return nil, err
+	}
+
+	return jobs, nil
+}
+
+//jobs/status
+func GetJobStatusCounts(db *sql.DB, startDay string, endDay string)([]Job, error){
+	var jobType *string
+	var count int
+	var status *string
+
+	query := `SELECT b.job_type, b.status, count(b.*) AS count
+            FROM (
+              SELECT a.id,a.status,
+                CASE WHEN array_length(a.types, 1) = 1 THEN a.types[1]
+              ELSE 'Mixed'
+           END AS job_type
+           FROM (
+               SELECT j.id, j.status, array_agg(DISTINCT t.name) AS types FROM jobs j
+               JOIN job_steps s ON j.id = s.job_id
+               JOIN job_types t ON s.job_type_id = t.id
+               WHERE j.start_date >= ($1 :: DATE) AND j.start_date <= ($2 :: DATE)
+               AND j.app_id != '1e8f719b-0452-4d39-a2f3-8714793ee3e6'
+               AND j.status in ('Completed', 'Failed', 'Canceled')
+               GROUP BY j.id
+           ) AS a
+          ) AS b
+         GROUP BY b.job_type, b.status
+         ORDER BY b.job_type;`
+
+	rows, err := db.Query(query, startDay, endDay)
+	if err != nil {
+		return nil, err;
+	}
+
+	defer rows.Close()
+
+	var jobs []Job
+
+	for i := 0; rows.Next(); i++ {
+		err := rows.Scan(&jobType, &status, &count)
+		if err != nil {
+			return nil, err;
+		}
+		jobs = append(jobs, Job{getStringValue(jobType), getStringValue(status), count})
```

Comment:
Noting my comment in the other function, this one's would be even nicer creating a `Job` at the start of the loop instead of scanning to intermediate variables.

I'm noticing the `getStringValue` and such here too though. By taking advantage of the zero values for types in your struct, it should be possible to not need that function (the zero value for a string is the empty string, for example). See https://tour.golang.org/basics/12 for a quick intro to zero values.

Reply:
Ah, right, I forgot about this. Coalescing in the database works -- `database/sql` also has its own somewhat-wonky way of dealing with it in the `sql.NullString` (https://golang.org/pkg/database/sql/#NullString) data type, but it can make it annoying to scan right into a struct, so it'd be a lot like the current thing anyway. Using `coalesce` is probably cleaner.